File: old_python/clifford/Clifford.py
from __future__ import annotations

import numpy as np
import pauli.Pauli as p
import clifford.clifford_from_matrix as cfm
import stabiliser_state.Check_Matrix as cm
import F2_helper.F2_helper as f2

class Clifford:

    @staticmethod
    def from_matrix(matrix: np.ndarray, assume_clifford : bool = False) -> Clifford: # TODO test
        clifford = cfm.Clifford_From_Matrix(matrix, allow_global_factor = True, assume_clifford = assume_clifford)

        if not clifford.is_clifford:
            raise ValueError('Matrix does not correspond to a Clifford')
        
        return clifford.get_clifford()

    def __init__(self, z_conjugates : list[p.Pauli], x_cojugates : list[p.Pauli], global_phase : complex = 1):
        self.number_qubits = len(z_conjugates)
        
        self.z_conjugates = z_conjugates
        self.x_conjugates = x_cojugates

        self.global_phase = global_phase

    def get_matrix(self) -> np.ndarray:
        size = 1 << self.number_qubits
        matrix = np.zeros((size, size), dtype = complex)
        
        first_col_check_matrix = cm.Check_Matrix(self.z_conjugates)
        matrix[:, 0] = first_col_check_matrix.get_state_vector() * self.global_phase

        old_col_index = 0

        for i in range(1, size):
            new_col_index = i ^ (i >> 1) # iterate through gray code so that we only flip one bit at a time, see https://www.geeksforgeeks.org/generate-n-bit-gray-codes/
            
            bit_flipped = f2.fast_log2(new_col_index ^ old_col_index)

            matrix[:, new_col_index] = self.x_conjugates[bit_flipped].multiply_vector(matrix[:, old_col_index])

            old_col_index = new_col_index

        return matrix

File: cpp/src/clifford/clifford.cpp
#include "clifford.h"
#include "util/f2_helper.h"
#include "stabiliser_state/check_matrix.h"
#include "stabiliser_state/stabiliser_state.h"

namespace fst
{
    Clifford::Clifford(const std::vector<Pauli> z_conjugates, const std::vector<Pauli> x_conjugates, const std::complex<float> global_phase )
        : z_conjugates(z_conjugates), x_conjugates(x_conjugates), global_phase(global_phase)
        {
            number_qubits = z_conjugates.size();
        }

    std::vector<std::vector<std::complex<float>>> Clifford::get_matrix() const
    {
        const std::size_t size = integral_pow_2(number_qubits);
        std::vector<std::vector<std::complex<float>>> transposed_matrix(size, std::vector<std::complex<float>> (size, 0) );

        Check_Matrix first_col_check_matrix(z_conjugates);
        Stabiliser_State state (first_col_check_matrix);
        state.global_phase = global_phase;

        transposed_matrix[0] = state.get_state_vector();

        std::size_t old_col_index = 0;

        for(std::size_t i = 1; i < size; i++)
        {
            // Iterate through the Gray code
            std::size_t new_col_index = i ^ (i >> 1);
            std::size_t bit_flipped = integral_log_2( old_col_index ^ new_col_index );

            transposed_matrix[new_col_index] = x_conjugates.at(bit_flipped).multiply_vector(transposed_matrix[old_col_index]);

            old_col_index = new_col_index;
        }

        std::vector<std::vector<std::complex<float>>> matrix(size, std::vector<std::complex<float>>(size, 0));
        
        for(std::size_t i = 0; i < size; i++)
        {
            for(std::size_t j = 0; j < size; j++)
            {
                matrix[i][j] = transposed_matrix[j][i];
            }
        }

        return matrix;
    }
}